core/bundle/changes/changes.go

```diff
@@ -1291,76 +1291,52 @@ func (cs *changeset) sorted() ([]Change, error) {
 	return sortedChanges, nil
 }
 
-// toposortFlatten performs a topographical flattened sort on the provided
+// toposortFlatten performs a stable topological sort on the provided
 // data.  dataOrder is a slice of the originally ordered changes. data is a
 // map of change to requirements.  To be idempotent, requirements must be in
 // the same order each time.  Use along with data to ensure that this method
 // is deterministic.
-//
-// Adapted from https://tylercipriani.com/blog/2017/09/13/topographical-sorting-in-golang/
-// Blog post: Topographical Sorting in Golang
-// Copyright © 2017 Tyler Cipriani
 func toposortFlatten(dataOrder []string, data map[string][]string) ([]string, error) {
 
-	// Create a map to handle degrees, all vertices start at 0.
+	// inDegree tracks the in-degree of each vertex `ch`
+	// i.e. the number of changes that must be done before `ch`.
 	inDegree := make(map[string]int, len(dataOrder))
-	for n := range data {
-		inDegree[n] = 0
-	}
-
-	// For each vertex, adjacent to "u", increment the degree.
-	for _, adjacent := range data {
-		for _, v := range adjacent {
-			inDegree[v]++
-		}
-	}
-
-	// Make a list of next, containing all vertex that have a degree
-	// of 0.
-	var next []string
-	for _, k := range dataOrder {
-		v := inDegree[k]
-		if v != 0 {
-			continue
+	// followers is the inverse of data: for each change `r`, followers[ch] is
+	// the list of all changes which require `r`.
+	followers := make(map[string][]string, len(dataOrder))
```

Comment:
Is this right? followers is given a size of len(dataOrder) but down below we expect there to be len(req) elements. But req is the data map values and could be variable and is unrelated to dataOrder? Not knowing the algorithm, are we sure that this is correct?

Reply:
Yep, followers  will only at most be the same size as data as it's just measuring the graph of requires statements.